### DISCUSSION TITLE
Scripting mono v2 - separate library into its own repository

### DISCUSSION THREAD

**Discussion by joelwurtz** [citizenfx/fivem#1982] [Engineering Discussion]
I saw there is a lot of recent changes into the C# script stuff with a new component going on.

I was wondering why keeping a lot of pure c# stuff into this repository and why not extracting some part of this library (at the exception of the event / tick / fivem integration) into it's own repository (installable as a nugget dependency) where it will be easier to upgrade / update and maybe have more contribution ?

**Comment by blattersturm** [comment]
It wouldn't be 'easier to update/upgrade' stuff that's out of the main repository: in fact, that would be *less* easy.

**Reply by joelwurtz** [reply]
I'm also a big advocate of the monorepository strategy as it's better to create feature that need to change various things on different components.

But maybe i'm wrong on how the C# part work, i know that actually each release create a new dll available on nugget, but there are mainly helper for developers as those dlls should not be copied into the module and the library will use the dll that are available on the client or server release of fivem.

I'm wondering if it would be possible, to split the dll into two : 

 * a main one for all the event handler / tick system / native link
 * a second one for others stuff (player / camera / world / animation / etc .... helpers)
 
The goal would be to let the first one in the client / server release (so no wrong link or weird stuff)
 And the second one will only be compiled / available in the resources directory like an external dependency 
 
Resources developer would then be able to embark new features from the second library without having to update their server release or wait that new features are available on the stable release for the client.

**Reply by blattersturm** [reply]
The new runtime actually splits relevant/'safe' parts into a library that's meant to be bundled with resources, though for the time being that's mainly for compatibility with different native declaration versions.

I didn't read the rest of your reply too much in depth, as it seems to be written under the assumption that there would still only be a stub library.

**Reply by thorium-cfx** [reply]
>But maybe i'm wrong on how the C# part work, i know that actually each release create a new dll available on nugget, but there are mainly helper for developers as those dlls should not be copied into the module and the library will use the dll that are available on the client or server release of fivem.

The assemblies on NuGet are purely reference assemblies, they actually contain no code and they will also not be loaded if you supply them.

> I'm wondering if it would be possible, to split the dll into two :
>
> a main one for all the event handler / tick system / native link
> a second one for others stuff (player / camera / world / animation / etc .... helpers)

Have a look at https://github.com/thorium-cfx/mono_v2_get_started, that should answer this question for v2 (v2 is inactive atm).

As to your other suggestions; we actually need to consider safe and unsafe code. Anything safe can just be supplied by you guys, which you can already do, anything unsafe needs to be supplied through here and everyone will need to wait for it.

**Comment by thorium-cfx** [comment]
@joelwurtz is there still more you'd like to discuss on this matter or can we continue closing this discussion as resolved?

**Reply by joelwurtz** [reply]
Sorry didn't close this, good for me thanks for your reply
